### ISSUE TITLE
Should whisk.invoke() return a Promise?

### ISSUE THREAD

**Issue by bjustin-ibm** [openwhisk/openwhisk#987]
I think the answer is definitely "yes" but not breaking existing action code makes it a little tricky. If I understand correctly, `whisk.invoke()` is currently always asynchronous and returns nothing.  If you specify a callback function as the `next` parameter, that callback will get called at some later time, depending on the value of the `blocking` parameter.

I propose that if `next` is left unspecified, then `whisk.invoke()` should return a Promise. If an error occurs, the promise is rejected with an error object (the same that would be passed to the callback, had `next` been specified). The promise is resolved on the activation in the same way that the callback would be called - as soon as the activation is created if `blocking` is false, or with the result if `blocking` is true.

### Non-blocking example
```javascript
whisk.invoke({
    name: 'myAction',
    blocking: false
  })
  .then(function (activation) {
      // activation completed successfully, but not necessarily finished
  })
  .catch(function (reason) {
      // activation was not created
      console.log('An error has occured ' + reason.error);
  });
```

### Blocking example
```javascript
whisk.invoke({
    name: 'myAction',
    blocking: true
  })
  .then(function (activation) {
      // activation completed successfully, activation contains the result
      console.log('Activation ' + activation.activationId + ' completed successfully and here is the result ' + activation.result);
  })
  .catch(function (reason) {
      console.log('An error has occured ' + reason.error);

      if(reason.activation) {
        console.log('Please check activation ' + reason.activation.activationId + ' for details.');
      } else {
        console.log('Failed to create activation.');
      }
  });
```

By doing it this way, existing actions do not need to change.  If an existing use of `whisk.invoke()` specifies a `next` parameter, the behavior is exactly the same - `whisk.invoke()` returns nothing, the callback is called as before. If they do not specify a `next` parameter, then their existing action code does not need to change because it already ignores what is returned (nothing) from calling `whisk.invoke()`.

**Comment by psuter** [comment]
Fully support the change, although I'd argue the examples should use `.catch` in the error handling part (easier to compose).

**Comment by bjustin-ibm** [comment]
@psuter Good point. I have updated the examples.

Additionally, when mocking up an implementation, it occurred to me that the activation is valuable even in the reject case. I have also updated the examples to demonstrate that the `reason` rejection object could have a combination of `error` and `activation` fields, depending on the value of `blocking`.

**Comment by rabbah** [comment]
If we use the openwhisk npm module for action invoke, how can we consolidate?

**Comment by csantanapr** [comment]
>Should whisk.invoke() return a Promise?

true

**Comment by rabbah** [comment]
While we pursue this now, we should also define the acceptance criteria to cut over from our own whisk.invoke to use `npm openwhisk` instead.

**Comment by csantanapr** [comment]
Yes, let's look over consolidating the npm module, and make it usable in both places, inside an action and from client side.
This will encourage us to take a look at the testing TODOs in the repo https://github.com/openwhisk/openwhisk-client-js/issues

**Comment by bjustin-ibm** [comment]
I agree that consolidating is a good idea, but should be handled in a separate issue.

**Comment by bjustin-ibm** [comment]
Resolved by #996
